test/test_type_promotion.py

```diff
@@ -345,17 +345,42 @@ def test_create_bool_tensors(self, device):
 
     @float_double_default_dtype
     def test_result_type(self, device):
-        self.assertEqual(torch.result_type(torch.tensor(1, dtype=torch.int, device=device), 1), torch.int)
-        self.assertEqual(torch.result_type(1, torch.tensor(1, dtype=torch.int, device=device)), torch.int)
+        # Tensor against tensor
+        for dtypea in torch.testing.get_all_dtypes():
+            for dtypeb in torch.testing.get_all_dtypes():
+                try:
+                    promoted_dtype = torch.promote_types(dtypea, dtypeb)
+                except RuntimeError:  # undefined promoted type
+                    with self.assertRaises(RuntimeError):
+                        torch.result_type(torch.tensor([0, 1], dtype=dtypea, device=device),
+                                          torch.tensor([1, 0], dtype=dtypeb, device=device))
+                    continue
+                self.assertEqual(torch.result_type(torch.tensor([0, 1], dtype=dtypea, device=device),
+                                                   torch.tensor([1, 0], dtype=dtypeb, device=device)), promoted_dtype)
+
+        # Tensor against scalar
+        for dtype in torch.testing.get_all_dtypes():
+            # int scalar
+            if dtype == torch.bool:
+                self.assertEqual(torch.result_type(torch.tensor([0, 1], dtype=dtype, device=device), 1), torch.int64)
+                self.assertEqual(torch.result_type(1, torch.tensor([0, 1], dtype=dtype, device=device)), torch.int64)
+            else:
+                self.assertEqual(torch.result_type(torch.tensor([0, 1], dtype=dtype, device=device), 1), dtype)
+                self.assertEqual(torch.result_type(1, torch.tensor([0, 1], dtype=dtype, device=device)), dtype)
+
+            # float scalar
+            if torch.testing.is_integral(dtype):
```

Comment:
These cases look good. How about adding a complex scalar case, too?

Reply:
I made a suggestion above that should naturally test for complex scalars, too, so you wouldn't have to add a special case.

Reply:
Please see my other comment and I'll address this accordingly.